### ISSUE TITLE
[REVIEW]: Datum: Creating and loading TFRecord datasets

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3446]
**Submitting author:** @n3011 (<a href="http://orcid.org/0000-0002-0059-2822">Mrinal Haloi</a>)
**Repository:** <a href="https://github.com/openAGI/datum" target ="_blank">https://github.com/openAGI/datum</a>
**Version:** 1.1.0
**Editor:** @jmschrei
**Reviewer:** @pmeier, @younglululu
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/5893594a099431c37c47276d6cb09119/status.svg)](https://joss.theoj.org/papers/5893594a099431c37c47276d6cb09119)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/5893594a099431c37c47276d6cb09119"><img src="https://joss.theoj.org/papers/5893594a099431c37c47276d6cb09119/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/5893594a099431c37c47276d6cb09119/status.svg)](https://joss.theoj.org/papers/5893594a099431c37c47276d6cb09119)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/openAGI/datum" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@pmeier & @younglululu, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @jmschrei know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/openAGI/datum">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@n3011) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/openAGI/datum">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@n3011) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @pmeier, @younglululu it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.08 s (880.0 files/s, 78984.6 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          52            702           1554           2714
Bourne Shell                     6            109             75            394
Markdown                         6            140              0            381
XML                              2              0              0            126
YAML                             4             12              0             76
-------------------------------------------------------------------------------
SUM:                            70            963           1629           3691
-------------------------------------------------------------------------------


Statistical information for the repository 'd30c3be86acc07c778b84698' was
gathered on 2021/07/02.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Ishant Mrinal Haloi             14          5103            133          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Ishant Mrinal Haloi        4970           97.4          0.2               18.91
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3446 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by jmschrei** [comment]
Howdy @pmeier and @younglululu

Thanks for agreeing to review this submission. 

The process for conducting a review is outlined above in the checklist from @whedon. Please check the boxes during your review to keep track, as well as making comments in this thread or opening issues in the repository itself to point out issues you encounter. Keep in mind that our aim is to improve the submission to the point where it is of high enough quality to be accepted, rather than to provide a yes/no decision, and so having a conversation with the authors is encouraged rather than providing a single review post at the end of the process.

Here are the review guidelines: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html
And here is a checklist, similar to above: https://joss.readthedocs.io/en/latest/review_checklist.html

Please let me know if you encounter any issues or need any help during the review process, and thanks for contributing your time to JOSS and the open source community!

**Comment by jmschrei** [comment]
@n3011 is there a paper for this submission? Is it possible to put it in a "paper" folder in the root directory?

**Comment by danielskatz** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03446/joss.03446/10.21105.joss.03446.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03446/joss.03446/10.21105.joss.03446.pdf) :page_facing_up: :point_left:

**Comment by jmschrei** [comment]
Thanks @danielskatz :)

**Comment by danielskatz** [comment]
@jmschrei - you can always look back at the prereview issue to see how the paper was generated when you aren't sure

**Comment by n3011** [comment]
> @n3011 is there a paper for this submission? Is it possible to put it in a "paper" folder in the root directory?

This issue is resolved, right?

**Comment by pmeier** [comment]
> This issue is resolved, right?

The paper is available on the [`joss` branch](https://github.com/openAGI/datum/tree/joss). 

That leaves the question which version / branch should be reviewed:
- `master`
- `joss`
- release `1.1.0`
?

**Comment by n3011** [comment]
`master` branch has the code of the recent release `1.2.0`. Please review `master` branch.

**Comment by pmeier** [comment]
This is my preliminary review based on commit [3878ec9](https://github.com/openAGI/datum/tree/3878ec9ac042e761626f9b3d3b9ecc40dd848317).

## Installation

Installation with `pip` it successful. I couldn't find any mention of supported Python versions. I've assumed 3.7, because this version is used in the development environment.

## Development

I couldn't find any documentation on how to run the linting and tests. I've assumed the commands extracted from the CI workflows. Both linting and tests are passing. For tests script reports that it will run the tests in parallel but doesn't. The test runner caught one warning pointing to `datum` that could be fixed.

## Documentation

I find the documentation insufficient. There are only two documents: the README and one other file in the `docs/` folder. The latter is not linked in the README and thus will only be found by looking through the source. Both contain some examples, but miss an API documentation. Without diving into the source, I didn't understand the interplay between the different classes.

Due to this, I was not able to try `datum` beyond the given examples. I'll hold my review of the functionality for now.

## Statement of Need / State of the field

The README and the paper claim that `datum` is able to reduce the amount of boilerplate code to write and read `TFRecords` datasets. Without significant effort it is impossible to validate that, since a side-by-side comparison is missing.

Furthermore, I couldn't find any mention of [`tensorflow-datasets`](https://www.tensorflow.org/datasets/). Just like `tensorflow`, this library is developed by Google and also provides functionalities to create `TFRecord` datasets. Thus, a comparison should probably use this as baseline rather than vanilla `tensorflow`.

**Comment by jmschrei** [comment]
Great, thanks for the review @pmeier. @younglululu, have you had a chance to start your review yet (see the checklist above)?

**Comment by whedon** [comment]
:wave: @younglululu, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @pmeier, please update us on how your review is going (this is an automated reminder).

**Comment by jmschrei** [comment]
@younglululu, @pmeier and @n3011, can you provide a brief update on how it's going? @n3011 are you getting useful feedback that is being incorporated into the submission? Thanks so much.

**Comment by younglululu** [comment]
I tried whatever I can to install the package, but failed regardless.

**Comment by pmeier** [comment]
@jmschrei I did my preliminary review in https://github.com/openjournals/joss-reviews/issues/3446#issuecomment-880581078. Since then there have been no commits, so I'm on hold for now.

**Comment by n3011** [comment]
@jmschrei  I have updated the project and the `paper.md` file to address comments from @pmeier . 

Package related changes:
1. Issue related to running tests in parallel is fixed and now in the master branch. 
2. API-doc strings are added. 

In the `joss` branch, `paper.md` and `paper.bib` file is updated to address the followings:
1. Information about TensorFlow-datasets. Tensorflow datasets is designed for publicly available datasets. To use it with custom datasets users still need to dataset-specific interface.
2. Included a para on the complexities involved in converting dataset to tfrecord format

**Comment by pmeier** [comment]
@n3011 

> To use it with custom datasets users still need to dataset-specific interface.

They have [extensive documentation](https://www.tensorflow.org/datasets/add_dataset) about how to add a custom dataset. From my current perspective they provide a superset of what `datum` provides. 

To convince me that this is not the case and `datum` indeed reduces the amount of boilerplate code needed to store and load a `TFRecord` dataset, I would like to see side-by-side comparison. That means, given the dataset structure you chose for `datum` what does the code look like to do it with `datum` and `tfds`.

**Comment by n3011** [comment]
@pmeier I will consider that as out of scope for this paper.

**Comment by jmschrei** [comment]
@n3011 thanks for responding to the feedback. I think that @pmeier's suggestion that you compare against TFDS is reasonable, and I would also like to see it. Please include such a comparison in your next revisions. Thanks!

**Comment by pmeier** [comment]
@n3011 

> I will consider that as out of scope for this paper.

Given that `datum`'s central claim is "Storing and saving `TFRecord` datasets is hard and tedious. `datum` helps with that.", providing evidence for this claim is very much in scope.

IMO, a side-by-side comparison is the easiest way to do that, but you are not bound to that. If you find another way to do such a comparison, I'm happy to review it.

**Comment by jmschrei** [comment]
Hi @n3011. Can you provide an estimate for how long it'll take for you to make the suggested changes? Thanks!

**Comment by younglululu** [comment]
This is my view of datum:

I followed the instruction and failed at the installation stage by using either pip install or conda install.

I created an issue regarding the failure (https://github.com/openAGI/datum/issues/8), and it was closed without solving the problem. 

There is no way I can review this paper without the successful installation.

**Comment by pmeier** [comment]
@younglululu openAGI/datum#8 should have been resolved by https://github.com/openAGI/datum/commit/b7f34ceedad48d03e50feb2b0e7a14ae353af0b5, which is included in the latest `datum==1.2.1` release. Have you retried the installation since then? If there is still some error, would you mind creating a new issue with the log output?

**Comment by jmschrei** [comment]
Hi @n3011. Can you please provide an update on how your revisions are going? If you are not interested in pursuing this, also please let us know. If we don't hear back from you, we'll have to reject the submission.

**Comment by arfon** [comment]
@n3011 – please let us know your intent with this submission. If we don't hear back from you in the next couple of weeks that you're planning on addressing the reviewer feedback then we will proceed to reject this submission.

**Comment by arfon** [comment]
@whedon reject

This submission is being rejected due to lack of activity from the authors. @pmeier, @younglululu, @jmschrei – thanks for your efforts here. Sorry this submission didn't make it through.

**Comment by whedon** [comment]
Paper rejected.
